core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/ldap/ReadOnlyLDAPUserStoreManager.java

```diff
@@ -1196,7 +1205,9 @@ private boolean bindAsUser(String userName, String dn, Object credentials) throw
                 log.debug("Authentication failed " + e);
                 log.debug("Clearing cache for DN: " + dn);
             }
-            userCache.remove(userName);
+            if (userName != null) {
+                userCache.remove(userName);
+            }
```

Comment:
We already have "boolean bindAsUser(String dn, String credentials)" function here. It would be better to use that with AuthenticationException thrown, and in the function in which you call this, catch AuthenticationException and remove user from cache. That would be the ideal fix. WDYT?